jena-fuseki2/jena-fuseki-core/src/main/java/org/apache/jena/fuseki/servlets/SPARQL_Query_Cache.java

```diff
@@ -0,0 +1,249 @@
+                    CacheAction cacheAction = new CacheAction(key,CacheAction.Type.READ_CACHE);
+                    sendResults(action, result, query.getPrologue(), cacheAction);
+                }
+
+        }
+        catch (QueryParseException ex) {
+            // Late stage static error (e.g. bad fixed Lucene query string).
+            ServletOps.errorBadRequest("Query parse error: \n" + queryString + "\n\r" + messageForQueryException(ex)) ;
+        }
+        catch (QueryCancelledException ex) {
+            // Additional counter information.
+            incCounter(action.getEndpoint().getCounters(), QueryTimeouts) ;
+            throw ex ;
+        } finally { action.endRead() ; }
+    }
+    /** Ship the results to the remote caller.
+     * @param action
+     * @param result
+     * @param qPrologue
+     */
+    protected void sendResults(HttpAction action, SPARQLResult result, Prologue qPrologue, CacheAction cacheAction) {
+        if ( result.isResultSet() )
+            ResponseResultSet.doResponseResultSet(action, result.getResultSet(), qPrologue, cacheAction) ;
+        else if ( result.isDataset() )
+            // CONSTRUCT is processed as a extended CONSTRUCT - result is a dataset.
+            ResponseDataset.doResponseDataset(action, result.getDataset());
+        else if ( result.isModel() )
+            // DESCRIBE results are models
+            ResponseDataset.doResponseModel(action, result.getModel());
+        else if ( result.isBoolean() )
+            ResponseResultSet.doResponseResultSet(action, result.getBooleanResult(), cacheAction) ;
+        else
+            ServletOps.errorOccurred("Unknown or invalid result type") ;
+    }
+
+    public static String generateKey(HttpAction action, Query query ,String queryString){
+        ResponseType responseType = null;
+        if(query.isAskType())
+            responseType = ResponseResultSet.getResponseType(action.getRequest(), DEF.rsOfferBoolean);
+        else if(query.isSelectType())
+            responseType = ResponseResultSet.getResponseType(action.getRequest(), DEF.rsOfferTable);
+        else if(query.isConstructType())
+            responseType = ResponseDataset.getResponseType(action.getRequest());
+        else if(query.isDescribeType())
+            responseType = ResponseDataset.getResponseType(action.getRequest());
+
+        return getKey(action, queryString, responseType);
+    }
+
+    private String formatForLog(Query query) {
+        IndentedLineBuffer out = new IndentedLineBuffer() ;
+        out.setFlatMode(true) ;
+        query.serialize(out) ;
+        return out.asString() ;
+    }
+
+    private static String getKey(HttpAction action, String queryString, ResponseType responseType) {
+        HttpServletRequest req = action.getRequest();
+        String uri = ActionLib.actionURI(req);
+        String dataSetUri = ActionLib.mapActionRequestToDataset(uri);
+        return dataSetUri + " " + queryString + " " + responseType.getContentType();
```

Comment:
It might be better to use the parsed `Query` for the caching key. Some texts for query could parse to the same Query (e.g. queries with different whitespace allocations).